src/lin_diffEq_HHL.jl

```diff
@@ -19,106 +19,91 @@ export QuEuler, QuLeapfrog, QuAB2, QuAB3, QuAB4
 """
 
 """
-    alg_lin_diffeq
+    AlgQuDiff
     * step - step for multistep method
     * α - coefficients for xₙ
     * β - coefficent for xₙ'
 """
-struct alg_lin_diffeq
-  step::Int
-  α::Array
-  β::Array
-  alg_lin_diffeq(step,α,β) = new(step,α,β)
+
+struct AlgQuDiff
+    step::Int
+    α::Array
+    β::Array
+    AlgQuDiff(step,α,β) = new(step,α,β)
 end
 
-function bval(alg::alg_lin_diffeq,t,h,g::Function)
-  b = zero(g(1))
-  for i = 1:(alg.step)
-    b += alg.β[i]*g(t-(i-1)*h)
-  end
-  b
+function bval(alg::AlgQuDiff,t,h,g::Function)
+    b = zero(g(1))
+    for i in 1:(alg.step)
+        b += alg.β[i]*g(t-(i-1)*h)
+    end
+    b
 end
 
-function Aval(alg::alg_lin_diffeq,t,h,g::Function)
-  I_mat = Matrix{Float64}(I, size(g(1)))
-  A = Matrix{Float64}(I, size(g(1)))
-  for i = 1:(alg.step)
-    A = [-1*(alg.α[i]*I_mat + h*alg.β[i]*g(t-(i-1)*h)) A]
-  end
-  A
+function aval(alg::AlgQuDiff,t,h,g::Function)
+    sz, = size(g(1))
+    A = Array{ComplexF64}(undef,sz,(alg.step + 1)*sz)
+    i_mat = Matrix{Float64}(I, size(g(1)))
+    A[1:sz,sz*(alg.step) + 1:sz*(alg.step + 1)] = i_mat
+    for i in 1:alg.step
+        A[1:sz,sz*(i - 1) + 1: sz*i] = -1*(alg.α[alg.step - i + 1]*i_mat + h*alg.β[alg.step - i + 1]*g(t - (alg.step - i)*h))
+    end
+    A
 end
 """
     Explicit Linear Multistep Methods
 """
-QuEuler = alg_lin_diffeq(1,[1],[1]);
-QuLeapfrog = alg_lin_diffeq(2,[0 1],[2 0])
-QuAB2 = alg_lin_diffeq(2,[1 0], [1.5 -0.5])
-QuAB3 = alg_lin_diffeq(3,[1 0 0], [23/12 -16/12 5/12])
-QuAB4 = alg_lin_diffeq(4,[1 0 0 0], [55/24 -59/24 37/24 -9/24])
-
-
-function prepare_init_state(tspan::Tuple,x::Vector,h::Float64,g::Function,alg::alg_lin_diffeq)
-  init_state = x;
-  N_t = Int(round(2*(tspan[2] - tspan[1])/h + 3))
-  b = similar(g(1))
-  for i = 1:N_t
-   if i < (N_t+1)/2 -1
-      b = bval(alg,h*(i-1) + tspan[1],h,g)
-      init_state = [init_state;h*b]
-    else
-      init_state = [init_state;zero(b)]
-   end
-
-  end
-  init_state = [init_state;zero(init_state)]
-  init_state
+QuEuler = AlgQuDiff(1,[1],[1]);
+QuLeapfrog = AlgQuDiff(2,[0 1],[2 0])
+QuAB2 = AlgQuDiff(2,[1 0], [1.5 -0.5])
+QuAB3 = AlgQuDiff(3,[1 0 0], [23/12 -16/12 5/12])
+QuAB4 = AlgQuDiff(4,[1 0 0 0], [55/24 -59/24 37/24 -9/24])
+
+
+function prepare_init_state(tspan::Tuple,x::Vector,h::Float64,g::Function,alg::AlgQuDiff)
+    N_t = Int(round(2*(tspan[2] - tspan[1])/h + 3))
+    sz, = size(g(1))
+    init_state = zeros(ComplexF64,2*(N_t + 1)*sz)
```

Comment:
I'm wondering if we could make this implementation more general? e.g it should work for different precision like `Float32`/ `Float64`, sometimes when we do not need high precision we could use low precision to make simulation faster.